File: csharp/1006.clumsy-factorial.cs
/*
 * @lc app=leetcode id=1006 lang=csharp
 *
 * [1006] Clumsy Factorial
 */

// @lc code=start
public class Solution {
    public int Clumsy(int N) {
        int ret = 0, acc = 0;
        for (int i = N; i > 0; --i) {
            int x = (N - i) % 4;
            if (x == 0) {
                if (i == N) {
                    acc = i;
                } else {
                    acc = -i;
                }
            } else if (x == 1) {
                acc *= i;
            } else if (x == 2) {
                ret += acc / i;
                acc = 0;
            } else {
                ret += i;
            }
        }
        return ret + acc;
    }
}
// @lc code=end

File: java/1006.clumsy-factorial.java
/*
 * @lc app=leetcode id=1006 lang=java
 *
 * [1006] Clumsy Factorial
 */

// @lc code=start
class Solution {
    public int clumsy(int n) {
        int ret = 0, acc = 1;
        for (int i = 0; i < n; ++i) {
            int d = n - i, r = i % 4;
            if (r < 2) {
                acc *= d;
            } else if (r == 2) {
                ret += acc / d;
                acc = -1;
            } else {
                ret += d;
            }
        }
        if (n % 4 < 3 && n % 4 > 0) {
            ret += acc;
        }
        return ret;
    }
}
// @lc code=end

